src/parser.rs

```diff
@@ -50,30 +53,30 @@ where Self: for<'a> TypeConstructor<'a> {
 
     /// This function can return an "Err" pseudo-element in order to support CSS2.1
     /// pseudo-elements.
-    fn parse_non_ts_pseudo_class(_context: &ParserContext,
+    fn parse_non_ts_pseudo_class(_context: &ParserContext<Self>,
                                  _name: &str)
         -> Result<Self::NonTSPseudoClass, ()> { Err(()) }
 
-    fn parse_non_ts_functional_pseudo_class(_context: &ParserContext,
+    fn parse_non_ts_functional_pseudo_class(_context: &ParserContext<Self>,
                                             _name: &str,
                                             _arguments: &mut Parser)
         -> Result<Self::NonTSPseudoClass, ()> { Err(()) }
 
     /// pseudo-elements
     type PseudoElement: Sized + PartialEq + Eq + Clone + Debug + Hash + MaybeHeapSizeOf;
-    fn parse_pseudo_element(_context: &ParserContext,
+    fn parse_pseudo_element(_context: &ParserContext<Self>,
                             _name: &str)
         -> Result<Self::PseudoElement, ()> { Err(()) }
 }
 
-pub struct ParserContext {
+pub struct ParserContext<Impl: ?Sized + SelectorImpl> {
```

Comment:
That was a syntax error, but I could make the same simplifications with the opposite: `trait SelectorImpl: Sized`. It’s less general, but I don’t think there’s really a reason to use an unsized type here, since we never need any value of that type.